src/NATS.Client.Core/Internal/SslStreamConnection.cs

```diff
@@ -80,12 +80,34 @@ public void SignalDisconnected(Exception exception)
         _waitForClosedSource.TrySetResult(exception);
     }
 
-    public async Task AuthenticateAsClientAsync(NatsUri uri)
+    public async Task AuthenticateAsClientAsync(NatsUri uri, TimeSpan timeout)
     {
         var options = SslClientAuthenticationOptions(uri);
         try
         {
-            await _sslStream.AuthenticateAsClientAsync(options).ConfigureAwait(false);
+            using var ctsTimer = new CancellationTokenSource();
+            using var ctsAuth = new CancellationTokenSource();
```

Comment:
why not use `ctsAuth = new CancellationTokenSource(timeout);` and avoid the 2nd task?